speakers-etl/src/load/publishesEntries.ts

```diff
@@ -0,0 +1,98 @@
+import { Semaphore } from 'async-toolbox'
+import { ParallelTransform } from 'async-toolbox/stream'
+import chalk from 'chalk'
+import { Asset, Entry, Environment } from 'contentful-management'
+
+export class PublishesEntries extends ParallelTransform {
+  private publishedAssetCount: number = 0
+  private publishedEntryCount: number = 0
+  private skippedCount: number = 0
+
+  constructor(public readonly environment: Environment, private readonly rateLimit: Semaphore) {
+    super({
+      objectMode: true,
+      // publish 1 entry at a time
+      maxParallelChunks: 1,
+      // allow 10 entries in the publish queue
+      highWaterMark: 10,
+    })
+  }
+
+  public async _transformAsync(chunk: Entry | Asset) {
+    if (chunk.isPublished() && !chunk.isUpdated()) {
+      this.push(chunk)
+      return
+    }
+
+    if (isAsset(chunk)) {
+      const publishedAsset: Asset | undefined = await this.publishAsset(chunk)
+      if (publishedAsset) {
+        this.push(publishedAsset)
+      }
+    } else {
+      const publishedEntry: Entry | undefined = await this.publishEntry(chunk)
+      if (publishedEntry) {
+        this.push(publishedEntry)
+      }
+    }
+    this.emit('status', `published ${this.publishedEntryCount} entries and ${this.publishedAssetCount} assets ` +
+      `(${this.skippedCount} skipped)`)
+  }
+
+  private async publishAsset(asset: Asset): Promise<Asset | undefined> {
+    try {
+      asset = await this.rateLimit.lock(() =>
+            asset.publish())
+    } catch (ex: any) {
+      if (!['VersionMismatch'].includes(ex.name)) {
+        throw ex
+      }
+
+      // version mismatch means that asset processing updated the asset while
+      // we were waiting and we need to reload it and try again
+      asset = await this.rateLimit.lock(() =>
+        this.environment.getAsset(asset.sys.id))
+      this.write(asset)
+      return
+    }
+    this.publishedAssetCount++
+    return asset
+  }
+
+  private async publishEntry(entry: Entry): Promise<Entry | undefined> {
+    try {
+      entry = await this.rateLimit.lock(() =>
+          entry.publish())
+    } catch (ex: any) {
+      if (!['UnresolvedLinks', 'InvalidEntry'].includes(ex.name)) {
+        throw ex
+      }
+
+      const data = JSON.parse(ex.message)
+      const err = data.details.errors[0]
+      if (!err) {
+        throw ex
+      }
+      if (err.name == 'unique' && err.path.join('.') == 'fields.slug.en-US') {
```

Comment:
is the logic for this allowing dups to be put in the system, but just remain unpublished?

Reply:
I think this is leftover copy-paste, I'll remove it